### DISCUSSION TITLE
Can`t find release notes for @graphql-codegen/typescript v4

### DISCUSSION THREAD

**Discussion by kyptov** [dotansimha/graphql-code-generator#9484] [General] [answered]
Is it a good idea changing Scalar type ID from `string` to `string | number`? I think it is better to left ID input as string for backward compability. And those who needs number - can change it.

I beleive using uuid for ID is a best choice and I think most of the devs thinks same way.

**Comment by simontaisne** [comment]
I'm also seeking the rationale behind this change. Meanwhile, I'm overriding the built-in `ID` scalar to a `string` type as described here https://the-guild.dev/graphql/codegen/plugins/typescript/typescript-resolvers#scalars.

```ts
const config: CodegenConfig = {
   // ...
   generates: {
     'path/to/file': {
       // plugins...
       config: {
         scalars: {
           ID: {
             input: 'string',
             output: 'string'
           },
         }
       },
     },
   },
};
 ```

**Comment by dotansimha** [comment, marked as answer]
Seems like GitHub cut off some of the release notes under Releases. 
You can find it here: 
https://github.com/dotansimha/graphql-code-generator/blob/master/packages/plugins/typescript/typescript/CHANGELOG.md#400

**Reply by simontaisne** [reply]
Thanks for sharing @dotansimha! Then I wonder, what's the best way to handle the ID within resolvers?

```ts
import { Resolvers } from './generated/types';

const resolvers: Resolvers = {
  Query: {
    whatever: (_, { id }) => {
      // id can be either string or number?
    },
  },
};
```

**Reply by n1ru4l** [reply]
@simontaisne In that case id should be typed as `string` (output). If the resolver type here is instead `number | string`, this would be considered a bug.

See the "Input Coercion" section here: https://spec.graphql.org/June2018/#sec-ID

**Reply by simontaisne** [reply]
Hey @n1ru4l, yes, the return type would be string, however, the input type is either `number` or `string`.

```ts
export type Scalars = {
  ID: { input: number | string; output: string; }
  ...
};
```

So how would you handle that at the resolver level? I guess we could convert the ID to a specific type. I just find it a bit cumbersome?

```tsx
const resolvers: Resolvers = {
  Query: {
    whatever: (_, { id }) => {
      const normalizedId: string = String(id); // Convert to string

      // Use the normalizedId in your resolver logic
    },
  },
};

```

**Reply by n1ru4l** [reply]
If the type if the argument is `string | number` there is a bug in graphql code generator (coercion input type instead of output type used here_. The arguments within the resolver function should be coerced to the output type.

**Reply by simontaisne** [reply]
> The arguments within the resolver function should be coerced to the output type.

So, following this [changelog](https://github.com/dotansimha/graphql-code-generator/blob/master/packages/plugins/typescript/typescript/CHANGELOG.md#major-changes) that would be correct then?

> Note that for `typescript-resolvers`, the type of ID needs to be inverted 🤷

```tsx
export type Scalars = {
  ID: {
    input: string;
    output: string | number;
  }
}
```

**Reply by eddeee888** [reply]
Hello @simontaisne 👋 

To generate types for the server, we'd have to change the default scalar types in codegen config:

```typescript
const config: CodegenConfig = {
   // ...
   generates: {
     'path/to/file': {
       // plugins...
       config: {
         scalars: {
           ID: { 
             input: 'string', 
             output: 'string | number' 
           },
         }
       },
     },
   },
};
```

This should correctly type coerced input in args. This also mean resolver functions can return `string | number` values that end up in the `parent` of subsequent resolvers.

---

Alternatively, you can make ID input/output string to preserve the previous behaviour (similar to what you suggested):

```typescript
const config: CodegenConfig = {
   // ...
   generates: {
     'path/to/file': {
       // plugins...
       config: {
         scalars: {
           ID: 'string'
         }
       },
     },
   },
};
```

**Reply by simontaisne** [reply]
Thanks @eddeee888, this confirms the current behavior then. Appreciate it 👍

**Reply by eddeee888** [reply]
Hi @simontaisne,

FYI, the default type for ID input/output has been reverted back to `string` in the latest patch released here: https://github.com/dotansimha/graphql-code-generator/pull/9501 🙂
